### PR TITLE
added PyPi coordinates

### PR DESCRIPTION
**Pull request by ror6ax**
To address https://github.com/strongbox/strongbox/issues/809

It's a semi - wip and is hence pretty bumpy. Ill be happy with any remarks that you might have.
One question is whether ID is a requirement to be had? It seems like it, together with version. I went with constructing it from most common unique elements. Please let me know if that is what's supposed to happen ;)

### CODE REVIEW

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/main/java/org/carlspring/strongbox/artifact/coordinates/PypiArtifactCoordinates.java [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+
+    public static final String PYTHON_TAG = "python_tag";
+
+    public static final String ABI_TAG = "abi_tag";
+
+    public static final String PLATFORM_TAG = "platform_tag";
+
+    public PypiArtifactCoordinates(String distribution,
+                                   String version,
+                                   String build_tag,
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/main/java/org/carlspring/strongbox/artifact/coordinates/PypiArtifactCoordinates.java]: Could you please follow the coding convention? Thanks!

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/main/java/org/carlspring/strongbox/artifact/coordinates/PypiArtifactCoordinates.java [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+                                   String python_tag,
+                                   String abi_tag,
+                                   String platform_tag){
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/main/java/org/carlspring/strongbox/artifact/coordinates/PypiArtifactCoordinates.java]: Cosmetic: Braces on new line, please.

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/main/java/org/carlspring/strongbox/artifact/coordinates/PyPiArtifactCoordinates.java [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+    public static final String ABI_TAG = "abiTag";
+
+    public static final String PLATFORM_TAG = "platformTag";
+
+    public PyPiArtifactCoordinates(String distribution,
+                                   String version,
+                                   String buildTag,
+                                   String pythonTag,
+                                   String abiTag,
+                                   String platform_tag)
```
[the remark this answers is not in the document]
Comment (ror6ax) [reply strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/main/java/org/carlspring/strongbox/artifact/coordinates/PyPiArtifactCoordinates.java]: @carlspring - can you please explain how to fix these? I have no idea why constant naming are an issue here and how Sonar wants them named.

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/main/java/org/carlspring/strongbox/artifact/coordinates/PyPiArtifactCoordinates.java [hunk 4]
```diff
@@ -0,0 +1,168 @@
+package org.carlspring.strongbox.artifact.coordinates;
+
+import java.util.Map;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import org.semver.Version;
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/main/java/org/carlspring/strongbox/artifact/coordinates/PyPiArtifactCoordinates.java]: Question: Is PyPi strictly semver-based, or does it also accept free-style versions?

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/test/java/org/carlspring/strongbox/artifact/PyPiArtifactCoordinateTest.java [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+
+/**
+ * @author Gregory Reshetniak
+ *
+ */
+public class PyPiArtifactCoordinateTest
+{
+
+    @Test
+    public void testArtifactPathToCoordinatesConversion()
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/test/java/org/carlspring/strongbox/artifact/PyPiArtifactCoordinateTest.java]: Could you please add some more tests for this based on real-world packages. Try and find some odd examples and check if it works.

**Comment by carlspring** [comment]
Perhaps this wasn't explicitly mentioned in the task's description, but there's a bit more to this... Along with the `ArtifactCoordinates` implementation, there needs to be a utility class which can convert paths to `ArtifactCoordinates` and vice-versa. (We've also been working on improving the [How To Implement-Your Own Repository Format](https://github.com/strongbox/strongbox/wiki/How-To-Implement-Your-Own-Repository-Format#artifact-coordinates) where this is described. Could you please have a look at the updated page and implement such a utility class and the tests for it?

Thanks! (And sorry, if the original task's description didn't mention this explicitly).

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/test/java/org/carlspring/strongbox/artifact/PyPiArtifactCoordinateTest.java (hunk 5 above)
Comment (ror6ax) [reply strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/test/java/org/carlspring/strongbox/artifact/PyPiArtifactCoordinateTest.java]: Sure, will add that.

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/main/java/org/carlspring/strongbox/artifact/coordinates/PyPiArtifactCoordinates.java (hunk 4 above)
Comment (ror6ax) [reply strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/main/java/org/carlspring/strongbox/artifact/coordinates/PyPiArtifactCoordinates.java]: I've never seen non-semver packages IRL, but the possibility is supported - https://www.python.org/dev/peps/pep-0440/#version-scheme

I'm afraid covering all of those variants would be not practical. We could make this data driven and look up how popular those naming  conventions are on PyPi...

Comment (carlspring) [reply strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/main/java/org/carlspring/strongbox/artifact/coordinates/PyPiArtifactCoordinates.java]: The problem is that not everybody follows the norms. For example, either Maven, nor NPM, or NuGet are strict on this. In theory,  both NPM and NuGet should be using Semver, but in real life, if you have a look at the packages available on these repositories, you'll see that a large part of the people aren't adhering to these rules. If we lock down the version validation to just Semver by default, this is likely to cause problems for many people. Also, could you please have a look at our page on [[Artifact Coordinae Validators]]?

Comment (ror6ax) [reply strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/main/java/org/carlspring/strongbox/artifact/coordinates/PyPiArtifactCoordinates.java]: With all possible version schemes and frankly nothing holding a person from making up their own, version format is frankly not parseable. Do we just treat is as a string then?

**Comment by carlspring** [comment]
@ror6ax ,

Would you like to continue investigating this, or are you no longer interested/able to do so?

**Comment by carlspring** [comment]
Thanks for your efforts, @ror6ax! However, as it needed some more work and you seemed to have abandoned this, another contributor implemented it.

Closing in favour of #1197.

**Pull request closed without merging**